openedx/core/djangoapps/content/learning_sequences/api/processors/enrollment.py

```diff
@@ -0,0 +1,41 @@
+"""
+Simple OutlineProcessor that removes items based on Enrollment and course visibility setting.
+"""
+from openedx.features.course_experience import COURSE_ENABLE_UNENROLLED_ACCESS_FLAG
+from student.models import CourseEnrollment
+
+from .base import OutlineProcessor
+
+
+class EnrollmentOutlineProcessor(OutlineProcessor):
+    """
+    Simple OutlineProcessor that removes items based on Enrollment and course visibility setting.
+    """
+    def usage_keys_to_remove(self, full_course_outline):
+        """
+        Return sequences/sections to be removed
+        """
+        # Public outlines and courses don't need to hide anything from the outline.
+        is_unenrolled_access_enabled = COURSE_ENABLE_UNENROLLED_ACCESS_FLAG.is_enabled(self.course_key)
+        is_course_outline_publicly_visible = full_course_outline.course_visibility in ["public", "public_outline"]
+
+        if is_unenrolled_access_enabled and is_course_outline_publicly_visible:
+            return frozenset()
+
+        # Students who are enrolled can see the full outline.
+        if CourseEnrollment.is_enrolled(self.user, self.course_key):
+            return frozenset()
+
+        # Otherwise remove everything:
+        seqs_to_remove = set(full_course_outline.sequences)
+        sections_to_remove = set(sec.usage_key for sec in full_course_outline.sections)
+
+        return frozenset(seqs_to_remove | sections_to_remove)
+
+    def inaccessible_sequences(self, full_course_outline):
+        """
+        Return a set/frozenset of Sequence UsageKeys that are not accessible.
+        """
+        if full_course_outline.course_visibility == "public_outline":
+            return set(full_course_outline.sequences)
+        return set()
```

Comment:
Nit: Please prefer `frozenset` over `set` when returning from these methods. Immutable data structures are a little more predictable, and we want to gently discourage calling code from mutating the results, since some of those patterns will bite us if we ever try to parallelize processor execution at some future point.